feature: go to implementations for dependency sources

kasiaMarek opened the pull request:
Go to implementation for dependency sources https://github.com/scalameta/metals-feature-requests/issues/119.
resolves: https://github.com/scalameta/metals/issues/1287
We index the type hierarchy using `ScalaTopLevelMtags`, only saving the name of the overridden symbol and its usage position, the actual symbol is resolved lazily.

Shortcomings:
 - won't work for renamed import

```
Benchmark                        Mode  Cnt  Score   Error  Units
MetalsBench.toplevelsScalaIndex    ss   10  0.332 ± 0.007   s/op
MetalsBench.typeHierarchyIndex     ss   10  0.331 ± 0.004   s/op
 ```
 
follow up: provide solution for handling renamed symbols

filipwiech commented:
This is awesome! Thank you for working on this, it's something I was missing when coming from the IntelliJ. :+1:

jkciesluk reviewed (changes requested):
That’s an amazing feature! I have few nitpicks

kasiaMarek commented on mtags/src/main/scala/scala/meta/internal/mtags/ScalaToplevelMtags.scala:
```diff
@@ -668,7 +752,9 @@ class ScalaToplevelMtags(
-  private def acceptTrivia(): Unit = {
+  private def acceptTrivia(): Option[Int] = {
+    var includedNewline = false
+    var ident = 0
```
😂 I know one stands for identifier and the other one for indentation but those look the same to me.

kasiaMarek commented on tests/unit/src/test/scala/tests/JarTopLevelsSuite.scala:
```diff
@@ -31,30 +34,79 @@ class JarTopLevelsSuite extends BaseTablesSuite {
+      List((filePath, "foo/Hello#", UnresolvedOverriddenSymbol("AnyVal", 32)))
+
+    var jarStmt: PreparedStatement = null
+    val jar =
+      try {
+        jarStmt = tables
+          .connect()
+          .prepareStatement(
+            s"insert into indexed_jar (md5) values (?)",
+            Statement.RETURN_GENERATED_KEYS,
+          )
+        jarStmt.setString(1, tables.jarSymbols.getMD5Digest(jar1))
+        jarStmt.executeUpdate()
+        val rs = jarStmt.getGeneratedKeys
+        rs.next()
+        rs.getInt("id")
+      } finally {
+        if (jarStmt != null) jarStmt.close()
+      }
```
No, that's not the same. We want to put into `indexed_jar` table only to `md5` column in this test. It's not something we do anymore but a backwards compatibility test.

kasiaMarek commented on tests/unit/src/test/scala/tests/ImplementationLspSuite.scala:
```diff
@@ -571,6 +573,62 @@ class ImplementationLspSuite extends BaseRangesSuite("implementation") {
+      |class MyException extends Excep@@tion
+      |""".stripMargin,
+    """|a/MyException#
+       |scala/ScalaReflectionException#
```
Yes

tgodzik replied:
Could we add it as the next step after this?

tgodzik reviewed (changes requested):
Should this work for Scala standard library? I am trying to find implementations of `Seq` but nothing is returned. Maybe you could add a test in ImplementationLspSuite?

tgodzik commented:
Wow, this is much faster now! 

I think there are still some smaller issues:
- no implementations are found for example for SeqFactory <- might be worth looking into that case
- getting an error even with a clean database:
```
org.h2.jdbc.JdbcBatchUpdateException: Unique index or primary key violation: "PUBLIC.PRIMARY_KEY_5 ON PUBLIC.TYPE_HIERARCHY(JAR, PATH, SYMBOL, PARENT_NAME) VALUES ( /* key:25106 */ 143, '/scala/tools/nsc/interactive/RangePositions.scala', 'scala/tools/nsc/interactive/RangePositions#', 'Positions')"; SQL statement:
insert into type_hierarchy (symbol, parent_name, path, jar, is_resolved) values (?, ?, ?, ?, ?) [23505-224]
	at org.h2.jdbc.JdbcPreparedStatement.executeBatch(JdbcPreparedStatement.java:1282)
	at scala.meta.internal.metals.JarTopLevels.putTypeHierarchyInfo(JarTopLevels.scala:210)
	at scala.meta.internal.metals.JarTopLevels.putJarIndexingInfo(JarTopLevels.scala:128)
	at scala.meta.internal.metals.Indexer.addSourceJarSymbols(Indexer.scala:639)
	at scala.meta.internal.metals.Indexer.$anonfun$indexDependencySources$5(Indexer.scala:469)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:903)
	at scala.meta.internal.metals.Indexer.$anonfun$indexDependencySources$1(Indexer.scala:460)
	at scala.meta.internal.metals.Indexer.$anonfun$indexDependencySources$1$adapted(Indexer.scala:459)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at scala.meta.internal.metals.Indexer.indexDependencySources(Indexer.scala:459)
	at scala.meta.internal.metals.Indexer.$anonfun$indexWorkspace$24(Indexer.scala:378)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.meta.internal.metals.TimerProvider.timedThunk(TimerProvider.scala:25)
	at scala.meta.internal.metals.Indexer.$anonfun$indexWorkspace$23(Indexer.scala:371)
	at scala.meta.internal.metals.Indexer.$anonfun$indexWorkspace$23$adapted(Indexer.scala:367)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at scala.meta.internal.metals.Indexer.indexWorkspace(Indexer.scala:367)
	at scala.meta.internal.metals.Indexer.$anonfun$profiledIndexWorkspace$2(Indexer.scala:159)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.meta.internal.metals.TimerProvider.timedThunk(TimerProvider.scala:25)
	at scala.meta.internal.metals.Indexer.$anonfun$profiledIndexWorkspace$1(Indexer.scala:159)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:687)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:842)

```

kasiaMarek commented:
> no implementations are found for example for SeqFactory <- might be worth looking into that case

fixed, it was Scala 3 find parents issue (I assume you had the issue with Scala 3, since I didn't with Scala 2)

> getting an error even with a clean database

Since we don't have positions anymore there might be duplications, I added distinct for that. I still deleted that primary key, I don't think it's useful for anything.

tgodzik reviewed (approved):
LGTM! Looks great and pretty fast too.

tgodzik commented:
Just a reminder, we will want to port the new interface method to the presentation compiler.

tgodzik commented:
Ach, an I noticed one more bug, but we can followup with it. `sealed traits` don't seem to work properly. `dotty.tools.repl.ParseResult` for example

tgodzik commented:
Actually, more things in `scala3-compiler_3-3.3.1-source.jar` don't work, might worth taking a look before merging

Z1kkurat commented:
This is amazing, I'm so excited about this feature, thanks a million for your effort! 🎉

kasiaMarek commented on tests/unit/src/test/scala/tests/ImplementationLspSuite.scala:
```diff
@@ -551,8 +571,9 @@ class ImplementationLspSuite extends BaseRangesSuite("implementation") {
+  // missing implementation of `info` for java pc
   check(
-    "java-implementation",
+    "java-implementation".ignore,
```
I ignored tests with java files. We need to implement `info` in java pc. Alternatively, we can force `Compilers` to use scala pc for `info` (at least for now), this should make `go to impl` work in java files that belong to scala build target.

kasiaMarek commented on tests/unit/src/test/scala/tests/ImplementationLspSuite.scala:
```diff
@@ -255,8 +254,9 @@ class ImplementationLspSuite extends BaseRangesSuite("implementation") {
+  // we currently don't collect information about overridden symbols in JavaMtags
   check(
-    "java-classes",
+    "java-classes".ignore,
```
To make this work we need `JavaMtags` (or `JavaTopLevelMatgs`) to return information about overridden symbols. Now we don't know that `RuntimeException` implements `Exception`, and only `MyException` will be found.

kasiaMarek commented on metals/src/main/scala/scala/meta/internal/search/SymbolHierarchyOps.scala:
```diff
@@ -0,0 +1,227 @@
+
+import org.eclipse.lsp4j.Location
+
+class SymbolHierarchyOps(
```
All those things are moved from `ImplementationProvider`. Those are not needed for go to implementation anymore but are still used by `Supermethods` and `RenameProvider`. I didn't want to try and remove those in this PR.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala:
```diff
@@ -289,215 +257,222 @@ final class ImplementationProvider(
-      allLocations.asScala.toSeq
-    }
+    val splitJobs =
+      symbolInfo(textDocument, source, dealised).flatMap { optSymbolInfo =>
```
```suggestion
      symbolInfo(textDocument, source, dealiased).flatMap { optSymbolInfo =>
```

tgodzik reviewed (commented):
I will still check out it locally to see if we didn't miss anything.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala:
```diff
@@ -375,6 +375,15 @@ final class BuildTargets private (
+      path: AbsolutePath,
+  ): Boolean = {
+    val possibleBuildTargets =
+      target :: buildTargetTransitiveDependencies(target).toList
```
```suggestion
      target :: buildTargetTransitiveDependencies(target).toSet
```
?

tgodzik commented on mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java:
```diff
@@ -179,6 +179,10 @@ public CompletableFuture<List<SyntheticDecoration>> syntheticDecorations(Synthet
 		return CompletableFuture.completedFuture(Collections.emptyList());
 	}
 
+	public CompletableFuture<List<PcSymbolInformation>> info(String symbol) {
```
When can it return multiple? Is there a known case?

kasiaMarek replied:
overloaded methods

tgodzik replied:
Though usually they have a different symbol

method()
method(+1)
etc.

so this should not repeat.

kasiaMarek replied:
Yes, but I want it to return all. When we search for implementations of methods, we find implementations of owner class/trait and only then implementations of method, needs to be this way since we index only class hierarchy. E.g. if `class A extends B` and we're looking for implementations of `B.method(+1)`, the searched method will be `A.method` but we don't know the disambiguator so I get them all to find the one I need. In `Compilers.scala` there is `info` and `infoAll`, the first one does exactly what you'd expect and returns only info about that exact symbol.

If this is too odd, we can add `overloadedSymbols` to `PcSymbolInformation` to learn all the overloaded and get info about each one by one.

kasiaMarek replied:
@tgodzik, would you prefer I change this?

tgodzik replied:
Could we get a full method symbol ? This should be possible using Mtags class. :thinking:

tgodzik reviewed (approved):
LGTM! Amazing work!
